File: sdk/typescript/src/utils/retry.ts
/**
 * Retry utilities for ALGO ZK Oracle SDK
 */

import { isRetryableError } from './errors';

export interface RetryOptions {
  maxAttempts: number;
  delay: number;
  backoffFactor: number;
  maxDelay: number;
}

const DEFAULT_RETRY_OPTIONS: RetryOptions = {
  maxAttempts: 3,
  delay: 1000, // 1 second
  backoffFactor: 2,
  maxDelay: 10000, // 10 seconds
};

export async function withRetry<T>(
  fn: () => Promise<T>,
  options: Partial<RetryOptions> = {}
): Promise<T> {
  const opts = { ...DEFAULT_RETRY_OPTIONS, ...options };
  let lastError: Error;

  for (let attempt = 1; attempt <= opts.maxAttempts; attempt++) {
    try {
      return await fn();
    } catch (error) {
      lastError = error as Error;

      if (attempt === opts.maxAttempts || !isRetryableError(lastError)) {
        throw lastError;
      }

      const delay = Math.min(
        opts.delay * Math.pow(opts.backoffFactor, attempt - 1),
        opts.maxDelay
      );

      await sleep(delay);
    }
  }

  throw lastError!;
}

function sleep(ms: number): Promise<void> {
  return new Promise(resolve => setTimeout(resolve, ms));
}

export function exponentialBackoff(attempt: number, baseDelay: number = 1000): number {
  return Math.min(baseDelay * Math.pow(2, attempt), 30000);
}

File: sdk/python/algo_zk_oracle/utils/retry.py
"""
Retry utilities for Apollon - ZK Oracle Oracle SDK
"""

import asyncio
import time
from typing import Callable, TypeVar, Optional, Any
from functools import wraps

from ..exceptions.errors import is_retryable_error

T = TypeVar('T')


class RetryConfig:
    """Configuration for retry behavior"""
    
    def __init__(
        self,
        max_attempts: int = 3,
        delay: float = 1.0,
        backoff_factor: float = 2.0,
        max_delay: float = 10.0,
        jitter: bool = True,
    ):
        self.max_attempts = max_attempts
        self.delay = delay
        self.backoff_factor = backoff_factor
        self.max_delay = max_delay
        self.jitter = jitter


def retry_sync(config: Optional[RetryConfig] = None):
    """Decorator for synchronous retry logic"""
    if config is None:
        config = RetryConfig()
    
    def decorator(func: Callable[..., T]) -> Callable[..., T]:
        @wraps(func)
        def wrapper(*args, **kwargs) -> T:
            last_error = None
            
            for attempt in range(1, config.max_attempts + 1):
                try:
                    return func(*args, **kwargs)
                except Exception as error:
                    last_error = error
                    
                    if attempt == config.max_attempts or not is_retryable_error(error):
                        raise error
                    
                    delay = calculate_delay(attempt, config)
                    time.sleep(delay)
            
            raise last_error
        
        return wrapper
    return decorator


def retry_async(config: Optional[RetryConfig] = None):
    """Decorator for asynchronous retry logic"""
    if config is None:
        config = RetryConfig()
    
    def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
        @wraps(func)
        async def wrapper(*args, **kwargs):
            last_error = None
            
            for attempt in range(1, config.max_attempts + 1):
                try:
                    return await func(*args, **kwargs)
                except Exception as error:
                    last_error = error
                    
                    if attempt == config.max_attempts or not is_retryable_error(error):
                        raise error
                    
                    delay = calculate_delay(attempt, config)
                    await asyncio.sleep(delay)
            
            raise last_error
        
        return wrapper
    return decorator


async def retry_async_func(
    func: Callable[..., Any],
    config: Optional[RetryConfig] = None,
    *args,
    **kwargs
):
    """Async function retry utility"""
    if config is None:
        config = RetryConfig()
    
    last_error = None
    
    for attempt in range(1, config.max_attempts + 1):
        try:
            return await func(*args, **kwargs)
        except Exception as error:
            last_error = error
            
            if attempt == config.max_attempts or not is_retryable_error(error):
                raise error
            
            delay = calculate_delay(attempt, config)
            await asyncio.sleep(delay)
    
    raise last_error


def retry_sync_func(
    func: Callable[..., T],
    config: Optional[RetryConfig] = None,
    *args,
    **kwargs
) -> T:
    """Sync function retry utility"""
    if config is None:
        config = RetryConfig()
    
    last_error = None
    
    for attempt in range(1, config.max_attempts + 1):
        try:
            return func(*args, **kwargs)
        except Exception as error:
            last_error = error
            
            if attempt == config.max_attempts or not is_retryable_error(error):
                raise error
            
            delay = calculate_delay(attempt, config)
            time.sleep(delay)
    
    raise last_error


def calculate_delay(attempt: int, config: RetryConfig) -> float:
    """Calculate delay for retry attempt with exponential backoff"""
    import random
    
    delay = config.delay * (config.backoff_factor ** (attempt - 1))
    delay = min(delay, config.max_delay)
    
    if config.jitter:
        # Add random jitter to avoid thundering herd
        delay *= (0.5 + random.random() * 0.5)
    
    return delay


def exponential_backoff(attempt: int, base_delay: float = 1.0) -> float:
    """Simple exponential backoff calculation"""
    return min(base_delay * (2 ** attempt), 30.0)